nvbench/detail/measure_cupti.cu

```diff
@@ -0,0 +1,288 @@
+  static constexpr const char *summary     = "L2 Cache Hit Rate (HW)";
+  static constexpr const char *hint        = "percentage";
+  static constexpr const char *short_name  = "L2HitRate";
+  static constexpr const char *description = "Hit rate at L2 cache.";
+  static constexpr double divider          = 100.0;
+
+  static bool is_collected(nvbench::state &m_state)
+  {
+    return m_state.is_l2_hit_rate_collected();
+  };
+};
+
+
+template <metric_ids id>
+void add_metric(nvbench::state &state,
+                std::vector<std::string> &metrics)
+{
+  if (metric_info<id>::is_collected(state))
+  {
+    metrics.emplace_back(metric_info<id>::metric_name);
+  }
+
+  add_metric<static_cast<metric_ids>(id + 1)>(state, metrics);
+}
+
+
+template <>
+void add_metric<metric_ids::count>(nvbench::state &m_state,
+                                   std::vector<std::string> &metrics)
+{}
+
+
+std::vector<std::string> add_metrics(nvbench::state &state)
+{
+  std::vector<std::string> metrics;
+  metrics.reserve(metric_ids::count);
+
+  add_metric<metric_ids::dram_peak_sustained_throughput>(state, metrics);
+  return metrics;
+}
+
+
+measure_cupti_base::measure_cupti_base(state &exec_state)
+  : m_state{exec_state}
+{
+  try
+  {
+    if (const std::optional<nvbench::device_info> &device = m_state.get_device();
+      device)
+    {
+      m_cupti = cupti_profiler(device, add_metrics(m_state));
+
+      if (!m_cupti->initialized())
+      {
+        m_cupti.reset();
+      }
+    }
+  }
+  catch (...)
+  {
+    m_cupti.reset();
+  }
+}
+
+
+void measure_cupti_base::check()
+{
+  if (!m_cupti)
+  {
+    if (auto printer_opt_ref = m_state.get_benchmark().get_printer();
+        printer_opt_ref.has_value())
+    {
+      auto &printer = printer_opt_ref.value().get();
+
+      printer.log(nvbench::log_level::fail,
+                  "CUPTI failed to collect metrics. "
+                  "Make sure you have permission to access NVIDIA GPU "
+                  "Performance Counters on the target device: "
+                  "https://developer.nvidia.com/"
+                  "nvidia-development-tools-solutions-err_nvgpuctrperm-"
+                  "permission-issue-performance-counters");
+    }
+  }
+
+  const auto device = m_state.get_device();
+  if (!device)
+  {
+    throw std::runtime_error(fmt::format("{}:{}: Device required for `cupti` "
+                                         "measurement.",
+                                         __FILE__,
+                                         __LINE__));
+  }
+  if (!device->is_active())
+  { // This means something went wrong higher up. Throw an error.
+    throw std::runtime_error(fmt::format("{}:{}: Internal error: Current "
+                                         "device is not active.",
+                                         __FILE__,
+                                         __LINE__));
+  }
+}
+
+
+template <metric_ids id>
+void gen_summary(std::size_t result_id,
```

Comment:
anon namespace, similar naming suggestion as the other metric helpers (`gen_summaries_impl`)